test/test-poll-oob.c

```diff
@@ -0,0 +1,200 @@
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#if !defined(_WIN32)
+
+#include "uv.h"
+#include "task.h"
+
+#include <errno.h>
+#include <sys/socket.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+#include <string.h>
+
+static uv_tcp_t server_handle;
+static uv_tcp_t client_handle;
+static uv_tcp_t peer_handle;
+static uv_poll_t poll_req[2];
+static uv_idle_t idle;
+static uv_os_fd_t client_fd;
+static uv_os_fd_t server_fd;
+static int ticks;
+static const int kMaxTicks = 10;
+static int cli_pr_check = 0;
+static int cli_rd_check = 0;
+static int srv_rd_check = 0;
+
+static int got_eagain(void) {
+  return errno == EAGAIN
+      || errno == EINPROGRESS
+#ifdef EWOULDBLOCK
+      || errno == EWOULDBLOCK
+#endif
+      ;
+}
+
+static void idle_cb(uv_idle_t* idle) {
+  uv_sleep(100);
+  if (++ticks < kMaxTicks)
+    return;
+
+  uv_poll_stop(&poll_req[0]);
+  uv_poll_stop(&poll_req[1]);
+  uv_close((uv_handle_t*) &server_handle, NULL);
+  uv_close((uv_handle_t*) &client_handle, NULL);
+  uv_close((uv_handle_t*) &peer_handle, NULL);
+  uv_close((uv_handle_t*) idle, NULL);
+}
+
+static void poll_cb(uv_poll_t *handle, int status, int events) {
+  char buffer[5];
+  int n;
+  int fd;
+
+  ASSERT(0 == uv_fileno((uv_handle_t*)handle, &fd));
+  memset(buffer, 0, 5);
+
+  if (events & UV_PRIORITIZED) {
+    n = recv(client_fd, &buffer, 5, MSG_OOB);
+    if (errno == EINVAL || errno == EINTR) {
+      return;
+    }
+    ASSERT(n > 0);
+    cli_pr_check = 1;
+  }
+  if (events & UV_READABLE) {
+    if (fd == client_fd) {
+      n = recv(client_fd, &buffer, 5, 0);
+      if (errno == EINVAL || errno == EINTR) {
+        return;
+      }
+      if (cli_rd_check == 1) {
+        ASSERT(strncmp(buffer, "world", n) == 0);
+        ASSERT(5 == n);
+        cli_rd_check = 2;
+      }
+      if (cli_rd_check == 0) {
+        ASSERT(n == 4);
+        ASSERT(strncmp(buffer, "hello", n) == 0);
+        cli_rd_check = 1;
+        do {
+          n = recv(client_fd, &buffer, 5, 0);
```

Comment:
So the OSX test shows that the suggested assertion fails the test, while we know it did succeed before. What are your ideas?